File: packages/expo-dual-camera/ios/DualCameraView.swift
import ExpoModulesCore
import UIKit
import AVFoundation

public class DualCameraView: ExpoView {

    // MARK: - Properties

    private var previewLayer: AVCaptureVideoPreviewLayer?
    private var side: String = "back"
    private var errorLabel: UILabel?

    // MARK: - Events

    let onCameraReady = EventDispatcher()
    let onMountError = EventDispatcher()

    // MARK: - Initialization

    public required init(appContext: AppContext? = nil) {
        super.init(appContext: appContext)
        backgroundColor = .black
    }

    @available(*, unavailable)
    public required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    // MARK: - Lifecycle

    public override func didMoveToWindow() {
        super.didMoveToWindow()
        NSObject.cancelPreviousPerformRequests(
            withTarget: self,
            selector: #selector(handleWindowChange),
            object: nil
        )
        perform(#selector(handleWindowChange), with: nil, afterDelay: 0)
    }

    @objc private func handleWindowChange() {
        if window != nil {
            DualCameraSessionManager.shared.register(self, side: side)
        } else {
            DualCameraSessionManager.shared.unregister(self)
        }
    }

    public override func layoutSubviews() {
        super.layoutSubviews()
        previewLayer?.frame = bounds
        errorLabel?.frame = bounds
    }

    // MARK: - Prop Setters

    func setSide(_ newSide: String) {
        guard newSide != side else { return }
        let wasAttached = window != nil
        if wasAttached { DualCameraSessionManager.shared.unregister(self) }
        side = newSide
        if wasAttached { DualCameraSessionManager.shared.register(self, side: side) }
    }

    func setLens(_ lens: String) {
        DualCameraSessionManager.shared.setBackLens(lens)
    }

    func setZoom(_ zoom: Double) {
        DualCameraSessionManager.shared.setZoom(side: side, normalizedZoom: zoom)
    }

    func setEnableTorch(_ enabled: Bool) {
        DualCameraSessionManager.shared.setTorch(enabled)
    }

    func setFlash(_ mode: String) {
        DualCameraSessionManager.shared.setFlash(mode)
    }

    func setMirror(_ mirror: Bool) {
        guard let connection = previewLayer?.connection else { return }
        if connection.isVideoMirroringSupported {
            connection.automaticallyAdjustsVideoMirroring = false
            connection.isVideoMirrored = mirror
        }
    }

    func setAutofocus(_ mode: String) {
        DualCameraSessionManager.shared.setAutofocus(side: side, mode: mode)
    }

    // MARK: - Preview Management

    func attachPreview(_ layer: AVCaptureVideoPreviewLayer) {
        previewLayer?.removeFromSuperlayer()
        layer.frame = bounds
        layer.videoGravity = .resizeAspectFill
        self.layer.insertSublayer(layer, at: 0)
        previewLayer = layer
        errorLabel?.isHidden = true
    }

    func detachPreview() {
        previewLayer?.removeFromSuperlayer()
        previewLayer = nil
    }

    // MARK: - Error Handling

    func showError(_ message: String) {
        if errorLabel == nil {
            let label = UILabel()
            label.backgroundColor = UIColor.black.withAlphaComponent(0.7)
            label.textColor = .white
            label.textAlignment = .center
            label.numberOfLines = 0
            label.font = .systemFont(ofSize: 14, weight: .medium)
            addSubview(label)
            errorLabel = label
        }
        errorLabel?.text = message
        errorLabel?.frame = bounds
        errorLabel?.isHidden = false

        onMountError(["message": message])
    }

    // MARK: - Session Callbacks

    func sessionDidStart() {
        onCameraReady([:])
    }

    // MARK: - Cleanup

    deinit {
        NSObject.cancelPreviousPerformRequests(withTarget: self)
        DualCameraSessionManager.shared.unregister(self)
    }
}

File: packages/expo-dual-camera/android/src/main/java/expo/modules/dualcamera/DualCameraView.kt
package expo.modules.dualcamera

import android.content.Context
import android.graphics.Color
import android.view.Gravity
import android.view.View
import android.widget.FrameLayout
import android.widget.TextView
import androidx.camera.view.PreviewView
import expo.modules.kotlin.viewevent.EventDispatcher

class DualCameraView(context: Context) : FrameLayout(context) {

    private val previewView: PreviewView
    private val errorView: TextView
    private var side: String = "back"

    val onCameraReady by EventDispatcher()
    val onMountError by EventDispatcher()

    init {
        previewView = PreviewView(context).apply {
            layoutParams = LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.MATCH_PARENT)
            implementationMode = PreviewView.ImplementationMode.PERFORMANCE
            scaleType = PreviewView.ScaleType.FILL_CENTER
            visibility = View.GONE
        }
        errorView = TextView(context).apply {
            setTextColor(Color.WHITE)
            setBackgroundColor(Color.BLACK)
            gravity = Gravity.CENTER
            visibility = View.GONE
            layoutParams = LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.MATCH_PARENT)
        }
        addView(previewView)
        addView(errorView)
    }

    // MARK: - Prop setters

    fun setSide(newSide: String) {
        if (newSide == side) return
        val wasAttached = isAttachedToWindow
        if (wasAttached) DualCameraSessionManager.unregister(this)
        side = newSide
        if (wasAttached) DualCameraSessionManager.register(this, side, context)
    }

    fun setLens(lens: String) {
        DualCameraSessionManager.setBackLens(lens, context)
    }

    fun setZoom(zoom: Double) {
        DualCameraSessionManager.setZoom(side, zoom.toFloat())
    }

    fun setEnableTorch(enabled: Boolean) {
        DualCameraSessionManager.setTorch(enabled)
    }

    fun setFlash(mode: String) {
        DualCameraSessionManager.setFlash(mode)
    }

    fun setMirror(mirror: Boolean) {
        previewView.scaleX = if (mirror) -1f else 1f
    }

    fun setAutofocus(mode: String) {
        DualCameraSessionManager.setAutofocus(side, mode)
    }

    // MARK: - Preview

    fun getPreviewView(): PreviewView = previewView

    fun showPreview() {
        previewView.visibility = View.VISIBLE
        errorView.visibility = View.GONE
    }

    fun showError(message: String) {
        errorView.text = message
        errorView.visibility = View.VISIBLE
        previewView.visibility = View.GONE
        onMountError(mapOf("message" to message))
    }

    fun sessionDidStart() {
        onCameraReady(mapOf<String, Any>())
    }

    // MARK: - Lifecycle

    override fun onAttachedToWindow() {
        super.onAttachedToWindow()
        DualCameraSessionManager.register(this, side, context)
    }

    override fun onDetachedFromWindow() {
        super.onDetachedFromWindow()
        DualCameraSessionManager.unregister(this)
    }
}
